### PR TITLE
Use configurable MaximumRecursion

### PR DESCRIPTION
**Pull request by tznind**
This PR addresses the issue where you need to give a large object with many child methods as a global variable but want to restrict the amount of up front reflection that happens (for performance).

https://github.com/NLua/NLua/issues/366

I have used System.DataTable in the test since it is a good complex class with self referencing properties.  

Setting `MaximumRecursion` to 0 on an instance of `Lua` results in only a single global being created and no reflection exploration happening (see TestMaximumRecursion).

### CODE REVIEW

File: tests/build/net45/NLuaTest.csproj [hunk 1]
```diff
@@ -45,6 +45,7 @@
       <HintPath>..\..\..\packages\NUnit.3.12.0\lib\net45\nunit.framework.dll</HintPath>
     </Reference>
     <Reference Include="System" />
```
Comment (viniciusjarina) [inline tests/build/net45/NLuaTest.csproj:47]: Why do we need the System.Data to test this? Can't we use a POCO object? NLua tests will run on Linux, Mac including iOS and tvOS.

Comment (tznind) [reply tests/build/net45/NLuaTest.csproj:47]: Good point.  I will fix this to use one of the test classes.

Comment (tznind) [reply tests/build/net45/NLuaTest.csproj]: So I changed to use the test class but Mac is still failing.  Do I need ` import ('mscorlib','System')` in DoString? I'm not sure what I am doing wrong.

File: tests/src/LuaTests.cs [hunk 2]
```diff
@@ -2795,6 +2795,32 @@ public void TestGuid()
...
+            using (Lua lua = new Lua())
+            {
+                lua.MaximumRecursion = maxRecursion;
+                lua.LoadCLRPackage();
+                lua["myTc"] = tc;
+                
+                if(maxRecursion == 0)
+                    Assert.AreEqual(1,lua.Globals.Count()); //register only the root reference
+                else
+                    Assert.Greater(lua.Globals.Count(),1); //many globals registered (all sub properties)
```
Comment (viniciusjarina) [inline tests/src/LuaTests.cs:2817]: NUnitLite doesn't have `Greater` you can just do `True (lua.Globals.Count(),1)`

Comment (tznind) [reply tests/src/LuaTests.cs]: Thanks for helping me with this PR.  Looks like there are some failing tests with `PlatformNotSupportedException` in NLua.CodeGeneration constructor now...

**Comment by tznind** [comment]
Ah looks like those PlatformNotSupported tests are also in your master branch: https://github.com/NLua/NLua/runs/463903787

**Pull request merged**